### PR TITLE
Generalize election management commands

### PR DESCRIPTION
**Pull request by z-bowen**
**Problem: CLI methods for election management need to be generalized away from `upsert-validator`**

## Solution

Rework the CLI methods to work with the more general `Election` base class

### CODE REVIEW

File: bigchaindb/commands/bigchaindb.py [hunk 1]
```diff
@@ -316,41 +315,38 @@ def create_parser():
...
-                                             help='Add/update/delete a validator.')
-
-    validator_subparser = validator_parser.add_subparsers(title='Action',
-                                                          dest='action')
-
-    new_election_parser = validator_subparser.add_parser('new',
-                                                         help='Calls a new election.')
+    # parser for managing elections
+    election_parser = subparsers.add_parser('election',
+                                            help='Add/update/delete a validator.')
```
Comment (ldmberman) [inline bigchaindb/commands/bigchaindb.py]: The help message needs an update.

File: docs/server/source/server-reference/bigchaindb-cli.md [hunk 2]
```diff
@@ -81,20 +81,25 @@ configuration file as documented under
...
+#### election new
 
-Create a new election which proposes a change to the validator set. An election can be used to add/update/remove a validator from the validator set.
+Create a new election which proposes a change to your BigChainDB network.
+
+There are multiple types of election, which each take different parameters. Below is a short description of each type of election, as well as their command line syntax and the return value.
+
+###### election new upsert-validator
+
+Call an election to add/update/remove a validator from the validator set.
```
Comment (kansi) [inline docs/server/source/server-reference/bigchaindb-cli.md]: ~Call~ Create/Generate

File: docs/server/source/server-reference/bigchaindb-cli.md [hunk 3]
```diff
@@ -119,13 +124,14 @@ If the command succeeds, it will create an election and return an `election_id`.
 **NOTE**: The election proposal consists of vote tokens allocated to each current validator as per their voting power. Validators then cast their votes to approve the change to the validator set by spending their vote tokens.
 
 
-#### upsert-validator approve
+#### election approve
+
+Approve an election by voting for it. The propsal generated by executing `bigchaindb election new ...` can approved by the validators using this command. The validator who is approving the proposal will spend all their votes i.e. if the validator has a network power of `10` then they will cast `10` votes for the proposal.
```
Comment (kansi) [inline docs/server/source/server-reference/bigchaindb-cli.md]: `can approved by ...`  -> `can be approved by ...`

File: docs/server/source/server-reference/bigchaindb-cli.md (hunk 2 above)
Comment (z-bowen) [reply docs/server/source/server-reference/bigchaindb-cli.md]: Fixed

File: docs/server/source/server-reference/bigchaindb-cli.md (hunk 3 above)
Comment (z-bowen) [reply docs/server/source/server-reference/bigchaindb-cli.md]: Fixed

File: bigchaindb/commands/bigchaindb.py [hunk 4]
```diff
@@ -317,41 +315,38 @@ def create_parser():
...
-    new_election_parser.add_argument('public_key',
-                                     help='Public key of the validator to be added/updated/removed.')
+    election_subparser = election_parser.add_subparsers(title='Action',
+                                                        dest='action')
 
-    new_election_parser.add_argument('power',
-                                     type=int,
-                                     help='The proposed power for the validator. '
-                                          'Setting to 0 will remove the validator.')
+    new_election_parser = election_subparser.add_parser('new',
```
Comment (kansi) [inline bigchaindb/commands/bigchaindb.py]: We dropped the [idea of using `new` keyword](https://github.com/bigchaindb/BEPs/pull/82/files#diff-ea17bcdbca54e1a9cd9a2705c65d9c64R128).

Comment (ldmberman) [inline bigchaindb/commands/bigchaindb.py]: @kansi has [suggested](https://github.com/bigchaindb/BEPs/pull/82#discussion_r216947470) to be more concise and just go with `bigchaindb election <type> ..` (no `new`). I like the idea.

File: bigchaindb/commands/bigchaindb.py [hunk 5]
```diff
@@ -317,41 +315,38 @@ def create_parser():
...
+    new_election_subparser = new_election_parser.add_subparsers(title='Election_Type',
+                                                                dest='election_type')
 
-    new_election_parser.add_argument('--private-key',
-                                     dest='sk',
-                                     help='Path to the private key of the election initiator.')
+    # Parser factory for each type of new election, so we get a bunch of commands that look like this:
+    # election new <some_election_type> <args>...
+    for name, data in elections.items():
+        help = data['help']
```
Comment (ldmberman) [inline bigchaindb/commands/bigchaindb.py]: `help` is a reserved keyword - I would just `.add_parser(name, data['help'])`.

File: bigchaindb/commands/bigchaindb.py (hunk 4 above)
Comment (ldmberman) [reply bigchaindb/commands/bigchaindb.py]: Discussed offline - removing it does not seem to be worth the effort so let's keep it.

File: bigchaindb/commands/bigchaindb.py (hunk 5 above)
Comment (z-bowen) [reply bigchaindb/commands/bigchaindb.py]: Fixed

File: docs/server/source/server-reference/bigchaindb-cli.md [hunk 6]
```diff
@@ -81,20 +81,25 @@ configuration file as documented under
 [Configuration Settings](configuration.html).
 
 
-## bigchaindb upsert-validator
+## bigchaindb election
 
-Manage elections to add, update, or remove a validator from the validators set. The upsert-validator subcommands implement [BEP-21](https://github.com/bigchaindb/BEPs/tree/master/21), please refer it for more details.
+Manage elections to manage the BigChainDB network. The specifics of the election process are defined in [BEP-21](https://github.com/bigchaindb/BEPs/tree/master/21), please refer it for more details.
```
[the remark this answers is not in the document]
Comment (z-bowen) [reply docs/server/source/server-reference/bigchaindb-cli.md]: Fixed

**Pull request merged**